### PR TITLE
Fixed #30382 -- Allowed specifying parent classes in force_insert of Model.save().

### PR DESCRIPTION
**Pull request by Akash-Kumar-Sen**
Fix for [issue#30382](https://code.djangoproject.com/ticket/30382)

Approach : 
- Discussion comment : https://github.com/django/django/pull/16830#discussion_r1198515059

### CODE REVIEW

**Comment by nessita** [comment]
I have started the review of this work but I'm having issues reproducing the original problem as described in the ticket. I have commented in there to hopefully obtain further details from the reporter.

**Comment by Akash-Kumar-Sen** [comment]
@nessita there might be another possible [issue #34549](https://code.djangoproject.com/ticket/34549), check my [comment](https://code.djangoproject.com/ticket/30382#comment:10) for more information on how to reproduce.

File: docs/releases/5.0.txt [hunk 1]
```diff
@@ -291,6 +291,19 @@ Models
...
+  to the ``force_insert`` parameter.
+
+* The new ``Model._meta.get_non_abstract_parent_list()`` method is a new
+  addition to the model's meta options. It returns all the non-abstract parent
+  classes of the model as a list.
+
+* ``Model.queryset.create()`` method now sets the ``force_insert`` parameter to
+  ``False`` when calling the :meth:`.Model.save` method if the model is
+  inherited from a non abstract parent class. This allows the creation of
+  inherited models without violating constraints.
```
Comment (felixxm) [inline docs/releases/5.0.txt:305]: These are not breaking changes but bugfixes and we don't document bugfixes in "feature" releases. Please revert changes to release notes.

File: docs/releases/5.0.txt [hunk 2]
```diff
@@ -276,7 +276,7 @@ Migrations
 Models
 ~~~~~~
 
-* The new ``create_defaults`` argument of :meth:`.QuerySet.update_or_create`
+* The new ``create_defaults`` argument of :meth:`.QuerySet.update_or_create`.
```
Comment (felixxm) [inline docs/releases/5.0.txt:279]: Please revert.

File: docs/releases/5.0.txt [hunk 3]
```diff
@@ -291,6 +291,19 @@ Models
...
 
+* :meth:`.Model.save` now passes the ``force_insert`` flag when saving models
+  on inherited parents. This ensures that all parent models are saved according
+  to the ``force_insert`` parameter.
+
+* The new ``Model._meta.get_non_abstract_parent_list()`` method is a new
+  addition to the model's meta options. It returns all the non-abstract parent
+  classes of the model as a list.
+
+* ``Model.queryset.create()`` method now sets the ``force_insert`` parameter to
```
Comment (felixxm) [inline docs/releases/5.0.txt:302]: Please move this to the `Backwards incompatible changes in 5.0` -> `Miscellaneous` section.

File: django/db/models/options.py [hunk 4]
```diff
@@ -711,6 +711,17 @@ def get_parent_list(self):
...
+    def get_non_abstract_parent_list(self):
+        """
+        Return all the non abstract ancestors of this model as a list.
+        """
+        parent_list = self.get_parent_list()
+        # Filter the list to exclude abstract models
+        concrete_parents = [
+            parent for parent in parent_list if not parent._meta.abstract
+        ]
+        return concrete_parents
```
Comment (felixxm) [inline django/db/models/options.py:723]: We don't use this list, so there is no need to create it, maybe
```suggestion
    def has_non_abstract_parent(self):
        return any(not parent._meta.abstract for parent in self.get_parent_list())
```
On the other hand to we need a separate helper for one-liner that is only used in one place?

File: django/db/models/base.py [hunk 5]
```diff
@@ -900,7 +902,7 @@ def save_base(
 
     save_base.alters_data = True
 
-    def _save_parents(self, cls, using, update_fields):
+    def _save_parents(self, cls, using, update_fields, force_insert=False):
```
Comment (shangxiao) [inline django/db/models/base.py:905]: My preference would be that we don't put a default for `force_insert` as it's a private method only called in one location outside any recursive calls.

File: django/db/models/options.py (hunk 4 above)
Comment (shangxiao) [reply django/db/models/options.py:723]: > On the other hand do we need a separate helper for one-liner that is only used in one place?

+1 too many abstractions can hinder maintenance

Comment (shangxiao) [reply django/db/models/options.py:723]: I like felix's `any()` variant too: Not only does it look more performant at a glance but it also reads better than generating a list then checking its length.

**Review by shangxiao — COMMENTED** [review]
Just added my 2¢

File: tests/model_inheritance/tests.py [hunk 6]
```diff
@@ -184,13 +184,11 @@ def b():
...
         for i, test in enumerate([a, b]):
-            with self.subTest(i=i), self.assertNumQueries(4), CaptureQueriesContext(
-                connection
-            ) as queries:
+            with self.subTest(i=i), CaptureQueriesContext(connection) as queries:
                 test()
                 for query in queries:
                     sql = query["sql"]
-                    self.assertIn("INSERT INTO", sql, sql)
+                    self.assertNotIn("UPDATE ", sql, sql)
```
Comment (shangxiao) [inline tests/model_inheritance/tests.py:191]: This changes the meaning of the test as it no longer proves that INSERTs were executed.

File: django/db/models/query.py [hunk 7]
```diff
@@ -647,7 +647,13 @@ def create(self, **kwargs):
         """
         obj = self.model(**kwargs)
         self._for_write = True
-        obj.save(force_insert=True, using=self.db)
+        force_insert = True
+        is_inherited = any(
+            not parent._meta.abstract for parent in self.model._meta.get_parent_list()
+        )
+        if is_inherited:
+            force_insert = False
```
Comment (shangxiao) [inline django/db/models/query.py:655]: Requires simplification :)

**Comment by nessita** [comment]
As mentioned in [the ticket](https://code.djangoproject.com/ticket/30382#comment:11), while this work does generate the expected queries for:

```python
ChildModel(id=1).save(force_insert=True)
```
For this code, an extra query is now executed:
```python
parent = ParentModel.objects.create(id=1)
ChildModel.objects.create(parentmodel_ptr=parent)
```

In `main`:
```
1. INSERT INTO "ticket_30382_parentmodel" ("id", "name") VALUES (1, '')
2. UPDATE "ticket_30382_parentmodel" SET "name" = '' WHERE "ticket_30382_parentmodel"."id" = 1
3. INSERT INTO "ticket_30382_childmodel" ("parentmodel_ptr_id") VALUES (1)
```

With this PR:
```
1. INSERT INTO "ticket_30382_parentmodel" ("id", "name") VALUES (1, '')
2. UPDATE "ticket_30382_parentmodel" SET "name" = '' WHERE "ticket_30382_parentmodel"."id" = 1
3. SELECT 1 AS "a" FROM "ticket_30382_childmodel" WHERE "ticket_30382_childmodel"."parentmodel_ptr_id" = 1 LIMIT 1
4. INSERT INTO "ticket_30382_childmodel" ("parentmodel_ptr_id") VALUES (1)
```

File: django/db/models/base.py [hunk 8]
```diff
@@ -811,11 +811,16 @@ def save(
             if loaded_fields:
                 update_fields = frozenset(loaded_fields)
 
+        calling_frame = inspect.currentframe().f_back
+        calling_function_name = calling_frame.f_code.co_name
+        for_create = calling_function_name == "create"
```
Comment (felixxm) [inline django/db/models/base.py:816]: This is very hacky. Unfortunately, we cannot accept such a solution because inspecting the stack frame is very expensive and would introduce performance regression, it's also unreadable, tricky, and error-prone.

I'm pretty sure there is an alternative solution for this, perhaps an extra layer/hook that will allow appropriate flags to be passed in a convenient way.

File: django/db/models/query.py [hunk 9]
```diff
@@ -647,6 +647,7 @@ def create(self, **kwargs):
         """
         obj = self.model(**kwargs)
         self._for_write = True
+        obj.for_create = True
```
Comment (felixxm) [inline django/db/models/query.py:650]: Unfortunately, we cannot do this as it would be backward incompatible for all models with `for_create` fields, methods, or properties.

File: django/db/models/query.py [hunk 10]
```diff
@@ -647,6 +647,7 @@ def create(self, **kwargs):
...
         obj.save(force_insert=True, using=self.db)
```
Comment (charettes) [inline django/db/models/query.py:651]: Have you considered changing the signature of `force_insert` to either allow passing a sentinel object to denote that only the current class should be inserted or maybe event a set of bases that should be forced to be inserted.

That would basically mean that `force_insert` signature could be changed from `bool` to `bool | set[Type[Model]]` and since a non empty set is still truthy we should maintain backward compatibilty to a reasonable extent.

Today `force_insert=True` means only force insert the current model, attempting to change its meaning to force insert all parents seems backward incompatible and the crux of the problem you are facing.

Maybe what we need to do here is
1. Allow bases to provided to `force_insert` as detailed above
2. Raise a deprecation warning when `force_insert=True` is used and MTI is involved to denote that it will mean force insert every bases in the future
3. Have `force_insert=True` default to `{self.__class__}` during the deprecation period.

Comment (Akash-Kumar-Sen) [reply django/db/models/query.py:651]: > 3. Have `force_insert=True` default to `{self.__class__}` during the deprecation period.

I think having `force_insert=False` will be a more appropriate default value here, as the current behavior is to not `force_insert` the parent when create is called.

File: docs/ref/models/instances.txt [hunk 11]
```diff
@@ -589,6 +589,24 @@ row. In these cases you can pass the ``force_insert=True`` or
...
+
+    In case of an inherited model the ``force_insert=True`` only forces the
+    insertion in the leaf model. ``force_insert=False`` is taken as default for the
+    parent models.
+
+    If it is required to force insert the parent models also, then consider passing
+    a tuple of Model classes that requires to perform force insertion. For
+    example::
+
+        Child(id=3, name="Child-3", value=10).save(force_insert=(GrandParent, Parent))
```
Comment (charettes) [inline docs/ref/models/instances.txt:602]: I think `(GrandParent, Parent)` is redundant with `GrandParent` as `issubclass(Parent, GrandParent)`?

File: docs/ref/models/instances.txt [hunk 12]
```diff
@@ -589,6 +589,24 @@ row. In these cases you can pass the ``force_insert=True`` or
...
+
+    ``force_insert=(models.Model,)`` can be used in order to perform forced
+    insertion of all the base classes. For example::
+
+        Child(id=1, name="Child-1", value=10).save(force_insert=(models.Model,))
```
Comment (charettes) [inline docs/ref/models/instances.txt:607]: Any thoughts about allowing passing a single model class instead of a singleton (e.g. `GrandParent` instead of `(GrandParent,)`).

I seems that it would be coherent with the signature of `issubclass` which allows either a `type` or `tuple[type]` ot be provided. Passing a tuple is also only useful for MTI models that are definitive multiple parents at once which is atypical scenario.

File: django/db/models/base.py [hunk 13]
```diff
@@ -660,6 +660,25 @@ def _set_pk_val(self, value):
                 setattr(self, parent_link.target_field.attname, value)
         return setattr(self, self._meta.pk.attname, value)
 
+    def _validate_force_insert(self, force_insert):
```
Comment (charettes) [inline django/db/models/base.py:663]: Not usage of `self` so this could be a `classmethod`

File: django/db/models/base.py [hunk 14]
```diff
@@ -660,6 +660,25 @@ def _set_pk_val(self, value):
...
+        if isinstance(force_insert, bool) or not force_insert:
+            return force_insert
+
+        if not isinstance(force_insert, tuple):
+            raise ValidationError("force_insert must be a bool or tuple")
+
+        if any(not issubclass(item, Model) for item in force_insert):
```
Comment (charettes) [inline django/db/models/base.py:670]: I think that the proper validation here is 

```suggestion
        if not all(isinstance(item, ModelBase) and issubclass(cls, item) for item in force_insert):
```

To ensure that `force_insert` doesn't include unrelated bases.

File: django/db/models/base.py [hunk 15]
```diff
@@ -660,6 +660,25 @@ def _set_pk_val(self, value):
...
+            raise ValidationError(
+                "All the items passed in force_insert must be a "
+                "subclass of models.Model"
+            )
+
+        if any(model == Model for model in force_insert):
+            model_tuple = (Model,)
+            return model_tuple
```
Comment (charettes) [inline django/db/models/base.py:678]: Is this a form of optimization to retrieve the _oldest_ common base?

File: django/db/models/base.py [hunk 16]
```diff
@@ -660,6 +660,25 @@ def _set_pk_val(self, value):
...
 
+    def _validate_force_insert(self, force_insert):
+        if isinstance(force_insert, bool) or not force_insert:
+            return force_insert
+
+        if not isinstance(force_insert, tuple):
+            raise ValidationError("force_insert must be a bool or tuple")
+
+        if any(not issubclass(item, Model) for item in force_insert):
+            raise ValidationError(
```
Comment (charettes) [inline django/db/models/base.py:671]: I think we'll want a `TypeError` here.

File: django/db/models/base.py [hunk 17]
```diff
@@ -660,6 +660,25 @@ def _set_pk_val(self, value):
                 setattr(self, parent_link.target_field.attname, value)
         return setattr(self, self._meta.pk.attname, value)
 
+    def _validate_force_insert(self, force_insert):
+        if isinstance(force_insert, bool) or not force_insert:
+            return force_insert
+
+        if not isinstance(force_insert, tuple):
+            raise ValidationError("force_insert must be a bool or tuple")
```
Comment (charettes) [inline django/db/models/base.py:668]: Ditto about `TypeError`

File: django/db/models/base.py [hunk 18]
```diff
@@ -660,6 +660,25 @@ def _set_pk_val(self, value):
                 setattr(self, parent_link.target_field.attname, value)
         return setattr(self, self._meta.pk.attname, value)
 
+    def _validate_force_insert(self, force_insert):
+        if isinstance(force_insert, bool) or not force_insert:
+            return force_insert
```
Comment (charettes) [inline django/db/models/base.py:665]: Could we switch the problem around and have `force_insert=True` turn into `(cls,)` and `force_insert=False` turn into `()`. Truthyness would be preserved and type checks for `force_insert` in `_save_parents` could be avoided.

**Review by charettes — COMMENTED** [review]
Thanks for the tweaks @Akash-Kumar-Sen!

I left some suggestions for improvements, happy to discuss them with you.

As for the new feature itself, it would be great to get some feedback from other members of the community. I believe that allowing an `issubclass` predicate of parents to be passed to `force_insert` is an elegant solution to the problem as explained on the mailing list but I've never actually needed this feature myself so it'd be great to have at least someone else's feedback.

Maybe we can get the original reporter of the issue to chime in here.

File: docs/ref/models/instances.txt (hunk 11 above)
Comment (Akash-Kumar-Sen) [reply docs/ref/models/instances.txt:602]: Yes, that's what I about to tell you as in Django one inherited model class cannot have more than one non-abstract parent class what is the point of passing a tuple, we can pass a single models.Model subclass.

Comment (charettes) [reply docs/ref/models/instances.txt:602]: Tuples, are only necessary when dealing with diamond inheritance.

Say that you have

```python
class A(models.Model):
   pass

class B(A):
   pass

class C(A):
   pass

class D(B, C):
   pass
```

It might be interesting to `force_insert` `B` and `C` while not `A`! Only in this case does passing a tuple is necessary as that's only doable by doing `d.save(force_insert=(B, C))`.

Comment (Akash-Kumar-Sen) [reply docs/ref/models/instances.txt:602]: Tried this but a_ptr willbe clashing in D, so it won't be a valid  inheritance I think.

Comment (Akash-Kumar-Sen) [reply docs/ref/models/instances.txt:602]: `
class GrandParent(models.Model):
    name = models.CharField(max_length=50, null=True)
    value = models.IntegerField(null=True)


class Parent(GrandParent):
    pass


class DiamondParent(GrandParent):
    pin_code = models.CharField(max_length=50, null=True)


class DiamondChild(Parent, DiamondParent):
    pass
`

Comment (charettes) [reply docs/ref/models/instances.txt:602]: This is expected, it happens because both intermediary parents must define an explicit `OneToOneField(GrandParent, primary_key=True, parent_link=True)` which distinct names.

File: django/db/models/base.py (hunk 15 above)
Comment (Akash-Kumar-Sen) [reply django/db/models/base.py:678]: Yes, It is, as it would reduce the number of checks later on in every recursive calls.
```
force_insert_parent = force_insert and any(
     issubclass(parent, model_class) for model_class in force_insert
)
```

**Comment by Akash-Kumar-Sen** [comment]
Thank you for the suggestions @charettes ! Updated the PR accordingly. Getting feedback from other members of the community would be great.

**Comment by felixxm** [comment]
@charettes Do you want to review it again? If not, I can try to push it forward.

File: django/db/models/base.py [hunk 19]
```diff
@@ -660,6 +660,34 @@ def _set_pk_val(self, value):
...
+            raise TypeError(
+                "All the items passed in force_insert must be a "
+                "subclass of models.Model"
+            )
+
+        if not all(issubclass(cls, item) for item in force_insert):
+            raise TypeError(
+                "The current model must be a subclass of"
+                "all the items passed in force_insert"
+            )
```
Comment (charettes) [inline django/db/models/base.py:687]: Feel like providing more details about which member is not a subclass of `cls` might be more helpful

```suggestion
        for base in force_insert:
            if not issubclass(cls, base):
                raise TypeError(
                    f"Invalid force_insert member. {base!r} must be a base of {cls!r}."
                )
```

File: django/db/models/base.py [hunk 20]
```diff
@@ -660,6 +660,34 @@ def _set_pk_val(self, value):
...
+            force_insert = (force_insert,)
+
+        if not isinstance(force_insert, tuple):
+            raise TypeError("force_insert must be a bool or tuple")
+
+        if not all(isinstance(item, ModelBase) for item in force_insert):
+            raise TypeError(
+                "All the items passed in force_insert must be a "
+                "subclass of models.Model"
+            )
```
Comment (charettes) [inline django/db/models/base.py:681]: Ditto

```suggestion
        for member in force_insert:
            if not isinstance(item, ModelBase):
                raise TypeError(
                    f"Invalid force_insert member. {member!r} must be a model subclass."
            )
```

File: docs/ref/models/instances.txt [hunk 21]
```diff
@@ -589,6 +589,26 @@ row. In these cases you can pass the ``force_insert=True`` or
...
 
+.. versionchanged:: 5.0
+
+    In case of an inherited model the ``force_insert=True`` only forces the
+    insertion in the leaf model. ``force_insert=False`` is taken as default for
+    the parent models.
+
+    If it is required to force insert the parent models also, then consider
+    passing a ``Model`` class or a tuple of ``Model`` classes that requires
+    to perform force insertion. For example::
```
Comment (charettes) [inline docs/ref/models/instances.txt:600]:
```suggestion
    To force the insertion of parent models as well you can pass the ancestor or a tuple
    of ancestors that should be inserted
```

File: tests/force_insert_update/tests.py [hunk 22]
```diff
@@ -66,3 +77,153 @@ def test_force_update_on_inherited_model_without_fields(self):
...
+                    force_insert=(models.Model,)
+                )
+
+        # gp E, p DNE
+        GrandParent(id=6, name="Grandparent-6", value=8).save()
+        with transaction.atomic():
+            with self.assertRaises(IntegrityError):
+                DiamondChild(id=6, name="Child-6", value=8).save(
+                    force_insert=(models.Model,)
+                )
```
Comment (charettes) [inline tests/force_insert_update/tests.py:206]: Are these tests failing due to #16903 and #16952? If it's the case then I think it's a strong argument to merge the other two first so we can have proper testing in place here.

**Review by charettes — COMMENTED** [review]
Nice work @Akash-Kumar-Sen!

The validation seem sane logic wise but I think the messaging could be improved.

@felixxm I'm not sure what the planed sequencing is here in terms of merging but I think that we should make sure diamond MTI insertion with a default pk works in the first place.

File: tests/force_insert_update/tests.py (hunk 22 above)
Comment (Akash-Kumar-Sen) [reply tests/force_insert_update/tests.py:206]: This is expected scenario, as the `GrandParent` already exists and we are trying to force insert that again it fails. This is the reason the force insert flag was not passed to `_save_parents` in the first place.

But it completely makes sense to pass #16952 and #16903 first before this one, as they are major bugs and keeps almost half the features of Django MTI unavailable.

In case of #16952 I am in completely blank as I am getting a completely unexpected error there. Your insights on this [comment](https://github.com/django/django/pull/16952#issuecomment-1581806139) willbe very much helpful.

File: tests/force_insert_update/models.py [hunk 23]
```diff
@@ -26,3 +26,26 @@ class SubCounter(Counter):
...
+
+
+class DiamondParent(GrandParent):
+    gp_ptr = models.OneToOneField(
+        GrandParent, primary_key=True, parent_link=True, on_delete=models.CASCADE
+    )
+
+
+class DiamondChild(Parent, DiamondParent):
+    pass
```
Comment (felixxm) [inline tests/force_insert_update/models.py:51]: Please try to reuse existing models (e.g. `Counter`, `SubCounter`) and minimize the number of new once.

**Review by felixxm — COMMENTED** [review]
@Akash-Kumar-Sen Thanks for updates :+1: I left more comments.

File: tests/force_insert_update/tests.py [hunk 24]
```diff
@@ -66,3 +77,153 @@ def test_force_update_on_inherited_model_without_fields(self):
         a.save()
         a.value = 2
         a.save(force_update=True)
+
+
+class ForceInsertInheritanceTests(TestCase):
+    def test_force_insert_on_grandparent_parent_doesnot_exists(self):
+        with self.assertNumQueries(3):
+            Child(id=1, name="Child-1", value=10).save(force_insert=(models.Model,))
```
[the remark this answers is not in the document]
Comment (felixxm) [reply tests/force_insert_update/tests.py:83]: This is not resolved.

File: django/db/models/base.py [hunk 25]
```diff
@@ -873,7 +898,11 @@ def save_base(
         with context_manager:
             parent_inserted = False
             if not raw:
-                parent_inserted = self._save_parents(cls, using, update_fields)
+                # validate force insert only if parents are inserted
+                force_insert = self._validate_force_insert(force_insert)
```
Comment (charettes) [inline django/db/models/base.py:902]: I think that's a compromise worth doing, that limits the possible extra cost of  ` _validate_force_insert` to MTI inserts 🎉

File: tests/force_insert_update/tests.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (charettes) [reply tests/force_insert_update/tests.py:83]: I think the idea is that from a third-party app perspective it might be useful to force the insertion of all bases of a Model subclasses independently  of its base chain.

In a certain sense `force_insert=(models.Model,)` is the initially requested feature of ticket-30382 to have `force_insert=True` force the insertion of all bases but since it would be backward incompatible to do so we document another way to achieve it.

[the remark this answers is not in the document]
Comment (Akash-Kumar-Sen) [reply tests/force_insert_update/tests.py:83]: And also in the current version of `_validate_force_insert()` it does not take any extra checks. So not supporting this will take extra checks now.

File: tests/force_insert_update/tests.py [hunk 26]
```diff
@@ -66,3 +75,153 @@ def test_force_update_on_inherited_model_without_fields(self):
...
+        with self.assertRaisesMessage(TypeError, expected_message_2):
+            Counter(id=1, name="Parent-1", value=8).save(force_insert=(TestCase,))
+
+        expected_message_3 = "force_insert must be a bool or tuple"
+        with self.assertRaisesMessage(TypeError, expected_message_3):
+            SubCounter(id=1, name="Parent-1", value=8).save(
+                force_insert=[
+                    Counter,
+                ]
+            )
```
Comment (felixxm) [inline tests/force_insert_update/tests.py:227]: I decided to reorganize new tests, some of them are redundant, some of them are unnecessary, some of them work without this patch. I moved the last group to a separate PR, check out #17028.

**Review by felixxm — COMMENTED** [review]
@Akash-Kumar-Sen Thanks :+1: I pushed edits to docs and reorganized tests.

@charettes Thanks for all reviews :1st_place_medal:

File: docs/ref/models/instances.txt [hunk 27]
```diff
@@ -589,13 +589,30 @@ row. In these cases you can pass the ``force_insert=True`` or
 Passing both parameters is an error: you cannot both insert *and* update at the
 same time!
 
+When using :ref:`multi-table inheritance <multi-table-inheritance>`, it's also
+possible to pass the ``force_insert`` with a tuple of parent classes to force
+an ``INSERT`` statements. For example::
```
Comment (charettes) [inline docs/ref/models/instances.txt:594]:
```suggestion
possible to provide a tuple of parents classes to ``force_insert`` in order to force 
`INSERT` statements for each base. For example::
```

File: docs/ref/models/instances.txt [hunk 28]
```diff
@@ -589,13 +589,30 @@ row. In these cases you can pass the ``force_insert=True`` or
...
+
+    Restaurant(pk=1, name="Bob's Cafe").save(force_insert=(Place,))
+
+    Restaurant(pk=1, name="Bob's Cafe", rating=4).save(force_insert=(Place, Rating))
+
+You can pass ``force_insert=(models.Model,)`` to force an ``INSERT`` statements
+for all parents. By default, ``force_insert=True`` forces insertion a new row
+only for the current model.
```
Comment (charettes) [inline docs/ref/models/instances.txt:602]:
```suggestion
You can pass ``force_insert=(models.Model,)`` to force an ``INSERT`` statement
for all parents. By default, ``force_insert=True`` only forces the insertion of a new
row for the current model.
```

File: docs/releases/5.0.txt [hunk 29]
```diff
@@ -335,6 +335,10 @@ Models
   :ref:`Choices classes <field-choices-enum-types>` directly instead of
   requiring expansion with the ``choices`` attribute.
 
+* The :ref:`force_insert <ref-models-force-insert>` argument of
+  :meth:`.Model.save` now allows specifying a tuple of parent classes to force
+  an ``INSERT`` statement.
```
Comment (charettes) [inline docs/releases/5.0.txt:340]:
```suggestion
  :meth:`.Model.save` now allows specifying a tuple of parent classes that must
  be forced to be inserted.
```

**Review by charettes — COMMENTED** [review]
Made a few suggestion to the docs, otherwise the change LGTM

**Comment by felixxm** [comment]
@charettes Thanks :+1: I pushed edits to docs.

**Review by charettes — APPROVED** [review]
